[CORE-14875] schema_registry: support metadata.properties

BenPope opened the pull request:
Added support for storing and retrieving metadata properties alongside the schema.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x

## Release Notes

### Features

* schema_registry: Support storing and retrieving Metadata Properties alongside the Schema.

BenPope committed:
sr/types: Prefer destructure for schemas

Whilst destructure isn't always convenient, the alternatives:

1. This looks like a little like use-after-move
```
    std::move(s).refs();
    std::move(s).raw();
```

2. When a new member is added, it's easy to forget to not move it.

Eliminate other rvalue-ref qualified members from:
* schema_definition
* subject_schema

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr/types: Add metadata to schemas

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr/storage: Add metadata serde

The format in the record is `<string, string>`.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr: Support registering and lookup of schema with metadata properties

The format that's accepted is:
`<string, string | integer | double | bool>`
Which is converted to:
`<string, string>`

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr: Support lookup of schema by id with metadata properties

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr: Add swagger for schemas with metadata properties

Signed-off-by: Ben Pope <ben@redpanda.com>

pgellert commented on src/v/pandaproxy/schema_registry/types.h:
```diff
@@ -168,6 +168,21 @@ class schema_definition {
           : schema_definition_iobuf{iobuf::from(sv)} {}
     };
     using references = chunked_vector<schema_reference>;
+    struct metadata {
```
nit: I think we could extract the metadata type out of `schema_definition` to the top-level, so that, for example, we could reuse it for the override/default metadata configs, and make code that takes metadata as an argument more concise (e.g., `const schema_definition::metadata&` -> `const metadata&`). But I'm also fine leaving this as is and refactoring later on demand.

BenPope replied:
Split out as `schema_metadata`

pgellert commented on src/v/pandaproxy/schema_registry/types.h:
```diff
@@ -168,6 +168,21 @@ class schema_definition {
     };
     using references = chunked_vector<schema_reference>;
+    struct metadata {
+        chunked_vector<std::pair<ss::sstring, ss::sstring>> properties;
```
What's the reasoning behind using a vector here? I'm wondering if it would be better to store this as an `absl::btree_map` for example to get the normalization behaviour out of the box. From the RFC:
> * Sort properties lexicographically by key.
> * For duplicate keys, the final occurrence wins.

Am I missing something or are these 2 normalization steps not yet implemented?

BenPope replied:
Changed to `btree_map`

pgellert commented on src/v/pandaproxy/schema_registry/requests/get_subject_versions_version.h:
```diff
@@ -54,6 +54,18 @@ void rjson_serialize(
+        }
+        w.EndObject();
+        w.EndObject();
+    }
```
nit: can we share this duplicated code?

BenPope replied:
split out into `rjson.h`

pgellert commented on tests/rptest/tests/schema_registry_test.py:
```diff
@@ -1729,6 +1729,86 @@ def test_post_subjects_subject_versions_metadata_ruleset(self):
+            subject=f"{topic}-key", data=schema_data
+        )
+        self.logger.debug(result_raw.content)
+        assert result_raw.status_code == requests.codes.ok
```
nit: `result_raw.content` only prints the response body I think, so if this ever fails on the status code, we'll have a harder time knowing what status code made it fail. You could use `self.assert_equal(...)` to surface the status code here.

BenPope replied:
The response is more fully logged in previous lines.

pgellert commented on src/v/pandaproxy/schema_registry/requests/post_subject_versions.h:
```diff
@@ -117,6 +136,7 @@ class post_subject_versions_request_handler
     bool Null() {
         switch (_state) {
         case state::metadata:
+        case state::metadata_properties:
```
Is this correct? Isn't this where we should return false to reject null values for metadata properties? From the RFC:
> Validation:
> * ...
> * null values are rejected.

BenPope replied:
Introduced a new state to distinguish this.

pgellert reviewed (commented):
Looks pretty good so far

BenPope committed:
treewide: Fix some missing includes

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr: Split out rjson package

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr: Move rjson_serialize schema_reference to rjson package

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr: modify POST /subjects/{subject}/versions

Return version, schemaType, references, and schema

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr: modify POST /subjects/{subject}

schemaType is always rendered

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr: modify GET /subjects/{subject}/versions/{version}

schemaType is always returned
deleted is now reterned

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr: Optimise sharded_store::has_schema

Pure refactor - no functional change.

1. Reduce cross-shard calls to subject shard
2. Search by greatest id and return first match

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented:
Changes in force-push:
* Addressed review comments
* Improved state json parse machines
* Improved responses to be more compatible with new behaviour

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--b86ec6e6-f133-45bd-8146-821a6dbf3366-->
<!--TEST RESULTS-->
<!--919fe807-3b41-4c7b-bd1f-e6bad203216b-->
<!--TEST RESULTS-->
<!--6b69b524-e6d7-4d71-9a1e-aa44f27badcc-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77877">build#77877</a></summary>


| test_class                 | test_method               | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                       |
|:---------------------------|:--------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionReassignmentsTest | test_reassignments_cancel | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77877#019b231f-e6fa-422a-9989-8e915ac80262 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0450, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1290, p1=0.2514, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionReassignmentsTest&test_method=test_reassignments_cancel |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77946">build#77946</a></summary>


| test_class          | test_method                                | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                 |
|:--------------------|:-------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CrashLoopChecksTest | test_cloud_storage_client_misconfiguration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77946#019b2741-7932-4c79-9506-9cf845377892 | FLAKY         | 2/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.9624, p0=0.9948, reject_threshold=0.0100. adj_baseline=0.9999, p1=0.0000, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CrashLoopChecksTest&test_method=test_cloud_storage_client_misconfiguration |
| CrashLoopChecksTest | test_cloud_storage_client_misconfiguration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77946#019b274b-cfe4-4a7d-ba48-57ab6db488da | FAIL          | 0/51     | Test PASSES after retries.[Inconclusive result after max retries](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#Case-F%3A-inconclusive-result-after-max-retries-(run-passes)%5BinlineExtension%5D)(baseline=0.9613, p0=0.1391, reject_threshold=0.0100. adj_baseline=0.9999, p1=1.0000, trust_threshold=0.5000)                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CrashLoopChecksTest&test_method=test_cloud_storage_client_misconfiguration |
| ScalingUpTest       | test_fast_node_addition                    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77946#019b274b-cfe0-4459-848e-d83154c06299 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0377, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1090, p1=0.3154, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_fast_node_addition                          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78532">build#78532</a></summary>


| test_class                         | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                   |
|:-----------------------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ScalingUpTest                      | test_scaling_up_with_recovered_topic | null             | integration | https://buildkite.com/redpanda/redpanda/builds/78532#019b8e55-4985-44ef-8a74-2aa4ea8a674c | FLAKY         | 19/21    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0253, p0=0.4012, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3917, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_scaling_up_with_recovered_topic               |
| CreateTopicReplicaDistributionTest | test_topic_aware_distribution        | null             | integration | https://buildkite.com/redpanda/redpanda/builds/78532#019b8e53-7458-4e3a-8ec5-ab7459f4d043 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CreateTopicReplicaDistributionTest&test_method=test_topic_aware_distribution |
| CreateTopicReplicaDistributionTest | test_topic_aware_distribution        | null             | integration | https://buildkite.com/redpanda/redpanda/builds/78532#019b8e55-497f-4302-b7cf-04964bf2896c | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CreateTopicReplicaDistributionTest&test_method=test_topic_aware_distribution |
</details>
<!--/body-->

BenPope committed:
sr: Support registering and lookup of schema with metadata properties - fallback

Schema should match when no metadata is provided.

The latest metadata is returned.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr/swagger: Extract schema_reference

Signed-off-by: Ben Pope <ben@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77946-->
<!--42ad026e-a28b-4598-999f-910101f0bc72-->
<!--title-->
## Retry command for [Build#77946](https://buildkite.com/redpanda/redpanda/builds/77946)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/crash_loop_checks_test.py::CrashLoopChecksTest.test_cloud_storage_client_misconfiguration
```
<!--/body-->

pgellert commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -637,11 +641,16 @@ post_subject_versions(server::request_t rq, server::reply_t rp) {
+        insert_result = co_await wr.write_subject_version(std::move(schema));
     }
 
+    auto r_schema = co_await st.get_schema_definition(insert_result.id);
```
We could probably reuse the earlier `auto s_id = co_await st.get_schema_id(schema.schema.def().share());` lookup to look up the schema definition (+ metadata?) to avoid the extra lookup here

IoannisRP replied:
`r_schema` and `schema` that was moved into `write_subject_version` should be the same, no?

BenPope replied:
> `r_schema` and `schema` that was moved into `write_subject_version` should be the same, no?

Yes, I think I can optimise this a little.

I was thinking that the advantage of going via subject is that it may be easier to handle `defautMetadata` and `overrideMetadata` if we support them on topic config at a later stage, but I'm not going via subject any more.

BenPope replied:
> We could probably reuse the earlier `auto s_id = co_await st.get_schema_id(schema.schema.def().share());` lookup to look up the schema definition (+ metadata?) to avoid the extra lookup here

It means returning all of the  schema across cores, just to throw all but one of them away. I've made the lookup conditional.

pgellert commented on src/v/pandaproxy/schema_registry/requests/get_subject_versions_version.h:
```diff
@@ -16,34 +16,32 @@
-    schema_id id;
-    schema_version version;
+struct get_subject_versions_version_response {
+    stored_schema stored_schema;
```
I like this introduction of endpoint-specific response structs. Should we try to create and share a standard `rjson_serialize` method for the `stored_schema` though to try to reuse as much as possible?

pgellert commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -178,14 +178,17 @@ struct json_schema_definition::impl {
     document_context ctx;
     ss::sstring name;
     schema_definition::references refs;
+    std::optional<schema_metadata> meta;
```
Does L204 need to include the (refs +) metadata now, as well as the equality check on L198?
```
std::ostream& operator<<(std::ostream& os, const json_schema_definition& def) {
```

pgellert commented on src/v/pandaproxy/schema_registry/types.h:
```diff
@@ -203,32 +221,39 @@ class schema_definition {
 
     const references& refs() const& { return _refs; }
 
+    const std::optional<schema_metadata>& meta() const& { return _meta; }
```
TIL: I didn't know `const&` is valid syntax here, not just `const`

IoannisRP replied:
My favorite part is that `const&&` is also valid :smile:

Is there any specific reason why these functions are `const&` instead of just `const`?
They should be equivalent (unless someone tries to move a const object :sweat_smile: ), its just that the `const` is more common.

BenPope replied:
> My favorite part is that `const&&` is also valid 😄
> 
> Is there any specific reason why these functions are `const&` instead of just `const`? They should be equivalent (unless someone tries to move a const object 😅 ), its just that the `const` is more common.

I didn't change it when I deleted the `&&` overload. I've changed it now since it looks odd.

pgellert commented on src/v/pandaproxy/schema_registry/storage.h:
```diff
@@ -538,6 +602,15 @@ class schema_value_handler final : public json::base_handler<Encoding> {
             return true;
         }
+        case state::metadata: {
+            _schema.metadata.emplace();
+            return true;
```
I'm not fully following the inconsistency here between the behaviour of metadata and metadata_properties.

Does this work for `{"metadata": null}`? Will we need to reset `_schema.metadata` at some point after we've emplaced here?

BenPope replied:
I've added tests and made a fix.

pgellert commented on src/v/pandaproxy/schema_registry/sharded_store.cc:
```diff
@@ -262,41 +262,34 @@ sharded_store::has_schema(subject_schema schema, include_deleted inc_del) {
-    struct match {
-        stored_schema sub_schema;
-        schema_id id;
-    };
-    std::optional<match> found;
-    for (auto ver : std::views::reverse(versions)) {
+    std::ranges::sort(versions, std::greater<>{}, &subject_version_entry::id);
+    for (auto entry : versions) {
```
I think we used to return a match for the maximal version if the id was the same. I think we should keep that behaviour, by sorting on a compound key of id + version.

BenPope replied:
> I think we used to return a match for the maximal version if the id was the same. I think we should keep that behaviour, by sorting on a compound key of id + version.

Done

pgellert reviewed (commented):
Looks pretty good to me, I only added some minor points

IoannisRP commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -510,7 +510,7 @@ post_subject(server::request_t rq, server::reply_t rp) {
 
     auto resp = ppj::rjson_serialize_iobuf(
-      post_subject_versions_version_response{
+      post_subject_response{
```
The commit message says:
> schemaType is always rendered

It was rendered before as well, wasn't it? The diff i see is that the `subject` has been added.

BenPope replied:
> The commit message says:
> 
> > schemaType is always rendered
> 
> It was rendered before as well, wasn't it? The diff i see is that the `subject` has been added.

It was conditionally rendered if it was not AVRO.

IoannisRP replied:
my bad. I was looking at the wrong type

IoannisRP commented on src/v/pandaproxy/schema_registry/types.h:
```diff
@@ -153,6 +154,18 @@ struct schema_reference {
+    operator==(const schema_metadata& lhs, const schema_metadata& rhs)
+      = default;
+
+    friend std::ostream& operator<<(std::ostream& os, const schema_metadata&);
```
are these still needed? I though we only used them to integrate with fmt. :thinking:

IoannisRP commented on src/v/pandaproxy/schema_registry/types.h:
```diff
@@ -153,6 +154,18 @@ struct schema_reference {
 };
 
+struct schema_metadata {
+    std::optional<absl::btree_map<ss::sstring, ss::sstring>> properties;
```
is this optional needed?

The metadata is `optional` in the `schema_definition`. Also, the map can be empty if there are no metadata.

BenPope replied:
I want to be able to represent each of these:
```json
{
  "metadata" : null
},
{
  "metadata": {
  }
},
{
  "metadata": {
    "properties": null
  }
},
{
  "metadata": {
    "properties": {}
  }
}
```
Later on there may be `ruleSet` and `metadata.tags`

IoannisRP commented on src/v/pandaproxy/schema_registry/types.h:
```diff
@@ -182,10 +195,15 @@ class schema_definition {
 
     template<typename T>
-    schema_definition(T&& def, schema_type type, references refs)
+    schema_definition(
+      T&& def,
+      schema_type type,
+      references refs,
+      std::optional<schema_metadata> meta)
```
nit: I think we are missing an explicit call to `_meta{}` in the constructor above, L192

BenPope replied:
> nit: I think we are missing an explicit call to `_meta{}` in the constructor above, L192

The default constructor of optional is sufficient, but yeah, I'll add it.

IoannisRP commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -178,14 +178,17 @@ struct json_schema_definition::impl {
       document_context ctx,
       std::string_view name,
-      schema_definition::references refs)
+      schema_definition::references refs,
+      std::optional<schema_metadata> meta)
```
I don't really like the asymmetry of json, where `refs` and `meta` are part of the pimpl, instead of the outer type, but that problem pre-exists this PR. Can be solved at a later time.

BenPope replied:
> I don't really like the asymmetry of json, where `refs` and `meta` are part of the pimpl, instead of the outer type, but that problem pre-exists this PR. Can be solved at a later time.

Yeah, ideally avro and proto aren't needed in `types.h`.

IoannisRP commented on src/v/wasm/schema_registry_module.cc:
```diff
@@ -84,7 +84,7 @@ read_encoded_schema_def(ffi::reader* r) {
         refs.emplace_back(name, subject(sub), schema_version(v));
     }
-    return {std::move(def), *type, std::move(refs)};
+    return {std::move(def), *type, std::move(refs), {}};
```
the metadata would be part of this message, wouldn't they? Is it ok for them to be ignored?

BenPope replied:
> the metadata would be part of this message, wouldn't they? Is it ok for them to be ignored?

WASM is out of scope.

IoannisRP commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -616,6 +621,28 @@ post_subject_versions(server::request_t rq, server::reply_t rp) {
+                      = std::move(r_schema).destructure();
+                    schema.schema = subject_schema{
+                      std::move(r_sub),
+                      {std::move(r_def), r_type, std::move(r_refs), meta}};
```
```suggestion
                      {std::move(r_def), r_type, std::move(r_refs), std::move(meta)}};
```

pgellert commented on src/v/pandaproxy/schema_registry/test/storage.cc:
```diff
@@ -198,3 +196,111 @@ BOOST_AUTO_TEST_CASE(test_storage_serde) {
+  "metadata": null,
+  "schema": "{\"type\":\"string\"}",
+  "deleted": true
+})"};
```
nit: a local schema factory method (e.g. `make_schema(std::optional<std::string_view> metadata)`) could reduce repetition and make this a bit easier to read.

BenPope replied:
> nit: a local schema factory method (e.g. `make_schema(std::optional<std::string_view> metadata)`) could reduce repetition and make this a bit easier to read.

Done

pgellert commented on src/v/pandaproxy/schema_registry/test/storage.cc:
```diff
@@ -198,3 +196,111 @@ BOOST_AUTO_TEST_CASE(test_storage_serde) {
+          ppj::rjson_serialize_str(val),
+          ::json::minify(empty_metadata_properties));
+        BOOST_CHECK(val.schema.def().meta().has_value());
+        BOOST_CHECK(val.schema.def().meta()->properties.has_value());
```
nit: should we add an assertion on the size of properties for these last 2 test cases?

BenPope replied:
> nit: should we add an assertion on the size of properties for these last 2 test cases?

Done

pgellert commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -616,6 +621,31 @@ post_subject_versions(server::request_t rq, server::reply_t rp) {
 
+        // If the provided schema lacks metadata, inherit from latest version
+        if (!schema.schema.def().meta().has_value()) {
+            try {
```
Is this done in import mode too? I would expect that it wouldn't be, because this introduces indeterminism (different results based on schema posting order) when importing schemas.

BenPope replied:
> Is this done in import mode too?

Fixed

pgellert commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -616,6 +621,31 @@ post_subject_versions(server::request_t rq, server::reply_t rp) {
+        // If the provided schema lacks metadata, inherit from latest version
+        if (!schema.schema.def().meta().has_value()) {
+            try {
+                auto latest = co_await st.get_subject_schema(
```
Would it make sense to populate the metadata earlier during the handler, before the `st.get_schema_id(...)` lookup on L566?

I am wondering if that way we could avoid the complexity of the asymmetric `matches` comparator:
```
/// Determine if this schema definition is equivalent to other; metadata is
/// only compared if this->meta().has_value().
inline bool matches(const schema_definition& other) const {
    bool meta_match = meta().has_value() ? meta() == other.meta() : true;
    return _def == other._def && _type == other._type
           && _refs == other._refs && meta_match;
}
```

pgellert reviewed (approved):
Lgtm
